# 데이터 분석 패키지 준비하기  
---

|패키지|기능|주요 함수|
|---|---|---|
|numpy|배열 계산, 반올림,버리기|floor,ceil,power|
|pandas|데이터프레임으로 데이터 입력,가공|describe,groupby|
|matplotlib|그래프 그리기|plot,hist,bar|  

외에도 많은 패키지들이 있음  

# 넘파이로 배열 다루기  
---

## 넘파이로 배열 정의하기

In [1]:
import numpy as np

In [2]:
a=np.array([[2,3],[5,2]])

In [3]:
a

array([[2, 3],
       [5, 2]])

|2|3|
|---|---|
|5|2|

## 배열 슬라이싱하기  
---

In [4]:
d=np.array([[1,2,3,4,5],[2,4,5,6,7],[5,7,8,9,9]])
d

array([[1, 2, 3, 4, 5],
       [2, 4, 5, 6, 7],
       [5, 7, 8, 9, 9]])

첫 번째 인덱스 : 행, 두 번째 인덱스 : 열

In [5]:
d[1][2]

5

In [6]:
d[1,2]

5

## 배열 크기 알아내기  
---

In [7]:
d=np.array([[2,3,4,5,6]])
d

array([[2, 3, 4, 5, 6]])

In [8]:
d.shape

(1, 5)

In [9]:
e=np.array([[1,2,3,4],[3,4,5,6]])
e

array([[1, 2, 3, 4],
       [3, 4, 5, 6]])

In [10]:
e.shape

(2, 4)

## 배열 원소 유형 확인하기  
---

In [12]:
d.dtype

dtype('int32')

|원소 유형|표기법|
|---|---|
|부호가 있는 정수|int(32,64)|
|부호가 없는 정수|unit(32,64)|
실수|float(32,64)|
|복소수|complex|
|참 거짓을 가지는 자료|bool|
|문자열|string|
|파이썬 오브젝트|object|
|유니코드|unicode|

## 배열 유형 바꾸기  
---

In [14]:
data=np.arange(1,5)

In [16]:
data.dtype

dtype('int32')

In [17]:
data.astype('float64')

array([1., 2., 3., 4.])

유형을 정수에서 실수로

## 넘파이 함수 알아보기  
---

- 0으로 이루어진 배열  

In [18]:
np.zeros((2,10)) #행:2,열:10

array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])

- 1으로 이루어진 배열  

In [19]:
np.ones((2,10))

array([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]])

- 연속형 정수 생성하기  

In [20]:
np.arange(2,10) #2이상10미만,1차원 배열

array([2, 3, 4, 5, 6, 7, 8, 9])

- 행과 열을 바꾸기 

In [21]:
a=np.ones((2,3))
a

array([[1., 1., 1.],
       [1., 1., 1.]])

In [22]:
b=np.transpose(a)
b

array([[1., 1.],
       [1., 1.],
       [1., 1.]])

## 배열의 사칙연산  
---

In [23]:
arr1=np.array([[2,3,4],[6,7,8]])
arr2=np.array([[12,23,34],[36,47,58]])

- 덧셈

In [24]:
arr1+arr2 #같은 자리끼리

array([[14, 26, 38],
       [42, 54, 66]])

- 곱셈

In [25]:
arr1*arr2 #같은 자리끼리

array([[ 24,  69, 136],
       [216, 329, 464]])

- 나눗셈

In [26]:
arr1/arr2 #같은 자리끼리

array([[0.16666667, 0.13043478, 0.11764706],
       [0.16666667, 0.14893617, 0.13793103]])

- 크기가 서로 다른 배열끼리 더하기

In [28]:
arr3=np.array([100,200,300])

In [30]:
arr1.shape

(2, 3)

In [31]:
arr3.shape

(3,)

In [32]:
arr1+arr3

array([[102, 203, 304],
       [106, 207, 308]])

- 브로드캐스팅이 되지 않는 경우

In [33]:
arr4=np.array([1,2,3,4,5,6,7,8,9,10])

In [34]:
arr4.shape

(10,)

In [35]:
arr1+arr4

ValueError: operands could not be broadcast together with shapes (2,3) (10,) 

 **행과 열이 모두 다르면 계산을 할 수 없음**

## 리스트와 배열의 차이점  
---

In [44]:
d=np.array([[1,2,3,4,5],[2,4,5,6,7],[5,7,8,9,9]]) # 3*5

In [45]:
d_list=[[1,2,3,4,5],[2,4,5,6,7],[5,7,8,9,9]]

In [46]:
d_list

[[1, 2, 3, 4, 5], [2, 4, 5, 6, 7], [5, 7, 8, 9, 9]]

In [47]:
type(d_list)

list

In [48]:
d_list[:2]=0

TypeError: can only assign an iterable

In [49]:
d[:2]=0

In [50]:
d

array([[0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0],
       [5, 7, 8, 9, 9]])

### 설문지 데이터 전처리  
---

5점 만점으로 된 설문 조사 결과

In [51]:
import os,usecsv
import numpy as np

In [53]:
os.chdir(r'C:\Users\phl02\Desktop\P\doit_python_programming')

In [54]:
quest=np.array(usecsv.switch(usecsv.opencsv('quest.csv')))

In [55]:
quest

array([[1., 2., 1., 2., 2.],
       [1., 3., 2., 3., 2.],
       [1., 4., 3., 3., 3.],
       [2., 5., 4., 4., 4.],
       [2., 5., 6., 2., 5.],
       [3., 6., 4., 2., 5.],
       [3., 5., 4., 1., 6.],
       [3., 5., 5., 1., 3.]])

In [56]:
quest[quest>5]

array([6., 6., 6.])

In [57]:
quest[quest>5]=5

In [58]:
quest

array([[1., 2., 1., 2., 2.],
       [1., 3., 2., 3., 2.],
       [1., 4., 3., 3., 3.],
       [2., 5., 4., 4., 4.],
       [2., 5., 5., 2., 5.],
       [3., 5., 4., 2., 5.],
       [3., 5., 4., 1., 5.],
       [3., 5., 5., 1., 3.]])

In [59]:
usecsv.writecsv('resultcsv.csv',list(quest))

# 넘파이로 사업성 분석하기  
---

<span style="color:limegreen">**사업성 분석이란?**</span>  
이 사업을 했을 때 이익이나 손해가 어느 정도 나올지 따져보는 것 

**경제적 타당성의 개념**  
이 사업이 우리에게 비용 대비 얼마나 많은 효용을 가져다 줄 수 있는지를 나타내는 개념  
- 순현재가치: 할인된 현금 흐름의 값을 모두 더한 값을 말한다.  
- 내부수익률: 순현재가치를 0으로 만드는 할인율을 말한다.

- 재무적 타당성의 세부 지표  

|이름|공식|설명|
|---|---|---|
|비용(C)|$\sum C$|사업을 만드는 데 들어가는 모든 비용|
|수입(B)|$\sum B$|사업을 하면서 발생하는 이익|
|현금 흐름(CF)|CF=B-C|수입에서 비용을 제외한 금액|
|연차(n)|-|총 사업 기간|
|할인율(r)|-|미래의 현금 흐름을 현재 기준으로 환산할 때 적용하는 비율|
|현금 흐름의 현재 가치(PV)|$PV=\frac{CF_k}{(1+r)^k}$|할인율을 적용한 k년차 현금 흐름의 현재 가치|
|순현재가치(NPV)|$NPV=\sum\frac{CF_k}{(1+r)^k}$|투자 안의 매년 현재 가치(PV)를 다 더한 값, 0보다 크면 사업성이 있다고 해석|

### 자본의 현재 가치 구하기  
---

In [60]:
import numpy as np

In [61]:
discount=0.05 #할인율 5%
cashflow=100 #현금 흐름 100억원

In [63]:
def pv(n):
    return(cashflow/((1+discount)**n))

In [64]:
pv(1) #1년이 지난 후 자본의 현재 가치

95.23809523809524

In [65]:
pv(2) #2년이 지난 후 자본의 현재 가치

90.70294784580499

In [66]:
for i in range(20):
    print(pv(i))

100.0
95.23809523809524
90.70294784580499
86.3837598531476
82.27024747918819
78.35261664684589
74.62153966366274
71.06813301301213
67.6839362028687
64.46089162177971
61.39132535407592
58.46792890864373
55.68374181775592
53.03213506452945
50.506795299551854
48.101709809096995
45.811152199140004
43.62966876108571
41.5520654867483
39.57339570166504


### 놀이공원 사업의 사업성 분석하기  
---

건설회사 사장이 놀이공원을 여는 것이 꿈이다. 건설사업이 성공해서 이제는 놀이공원에 한번 도전해도 괜찮을 것 같다.  
건설회사 사장님이 모은 돈과 지금 가진 신용으로 이 놀이공원을 지을 수 있을지, 지었을 때 이익이 날지 손해가 날지 알고 싶다.

&#128528; 여기서는 실습의 편의를 위해 공사를 시작하는 시점을 1년 차로 가정, 일반적으로 운영 수입이 발생하는 시점을 기준  
&#128528; 가상의 사업을 분석하는 과정에서 분석 기간은 사업에 따라서 달라질 수 있다.

![KakaoTalk_20220612_152617234](https://user-images.githubusercontent.com/105963819/173220123-a68814cd-29d2-414b-bbbc-ca964de58565.jpg)


- 세부 지표 값 구하기

In [67]:
loss=[-750,-250] #1,2년 차
profit=[100]*18 #3년 차부터 
cf=loss+profit
cf

[-750,
 -250,
 100,
 100,
 100,
 100,
 100,
 100,
 100,
 100,
 100,
 100,
 100,
 100,
 100,
 100,
 100,
 100,
 100,
 100]

In [68]:
len(cf)

20

In [69]:
cashflow=np.array(cf)

- 순현재가치와 내부수익률 구하기

In [70]:
npv=np.npv(0.045,cashflow)

<ipython-input-70-869079e5cd3a>:1: DeprecationWarning: numpy.npv is deprecated and will be removed from NumPy 1.20. Use numpy_financial.npv instead (https://pypi.org/project/numpy-financial/).
  npv=np.npv(0.045,cashflow)


In [71]:
npv

174.4011295118405

In [72]:
irr=np.irr(cashflow)

<ipython-input-72-7c34e3a77d23>:1: DeprecationWarning: numpy.irr is deprecated and will be removed from NumPy 1.20. Use numpy_financial.irr instead (https://pypi.org/project/numpy-financial/).
  irr=np.irr(cashflow)


In [73]:
irr

0.06434936937255586

- 핵심 지표 값 해석하기

순현재가치는 약 174.4억이고, 내부수익률은 6.4%정도가 나왔다.  
사실 재무적 타당성은 순현재가치나 내부수익률을 구하는 것보다 해석하는 일이 더 중요  
가정한 할인율은 4.5%였고, 6.4%는 4.5%보다 높기 때문에 수익성이 있다고 볼 수 있다.  
하지만 이자를 해마다 10%씩 내야 한다고 가정하면 다시 생각해야 한다.  
18년 동안 운영한다고 가정했을 때 내부수익룰이 6.4%이므로 10%보다 더 작다.  
현실에서는 할인율뿐만 아니라 여러 요소를 고려해서 분석해야 한다.

프로젝트의 가치는 174억으로서 이 정도 수익을 올릴 수 있다면. 이 사업은 순현재가치를 봤을 때도 사업성이 있는 것으로 판단

# 판다스로 데이터프레임 만들기  
---